flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java

```diff
@@ -0,0 +1,382 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.aggregate;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.dataview.MapView;
+import org.apache.flink.table.data.DecimalData;
+import org.apache.flink.table.data.DecimalDataUtils;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.apache.commons.math3.util.Pair;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.types.utils.DataTypeUtils.toInternalDataType;
+
+/** Built-in PERCENTILE aggregate function. */
+@Internal
+public abstract class PercentileAggFunction<T>
+        extends BuiltInAggregateFunction<T, PercentileAggFunction.PercentileAccumulator> {
+
+    protected final transient DataType valueType;
+    protected final transient DataType frequencyType;
+
+    public PercentileAggFunction(LogicalType inputType, LogicalType frequencyType) {
+        this.valueType = toInternalDataType(inputType);
+        this.frequencyType = frequencyType == null ? null : toInternalDataType(frequencyType);
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Accumulator
+    // --------------------------------------------------------------------------------------------
+
+    /** Accumulator for PERCENTILE. */
+    public static class PercentileAccumulator {
+
+        public double[] percentages;
+        public long totalCount;
+        public MapView<Double, Long> valueCount;
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            PercentileAccumulator that = (PercentileAccumulator) o;
+            return Arrays.equals(percentages, that.percentages)
+                    && totalCount == that.totalCount
+                    && Objects.equals(valueCount, that.valueCount);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(Arrays.hashCode(percentages), totalCount, valueCount.hashCode());
+        }
+
+        public Double[] getValue() {
+            List<Pair<Double, Integer>> sortedPercentages =
```

Comment:
Avoid Java Streams (https://flink.apache.org/zh/how-to-contribute/code-style-and-quality-java/#java-streams)